### PR TITLE
[JENKINS-72105] Fix scriptler script's parameter

### PR DESCRIPTION
**Pull request by limejuny**
### Detail

 - [JENKINS-72105](https://issues.jenkins.io/browse/JENKINS-72105)
 - When there is no `parameters`(deprecated param) in `scriptlerScript`, `parameters` is set with `Collections.emptyMap()`. And it is not null, so `parameters` and `parameters` in `ScriptlerBuilder` in `initializeFromDeprecatedProperties` is empty map and so parameter can't created.
 - I tested with snapshot version of scriptler-plugin, but I don't know the new version of scriptler dependency is mandatory. So I create commit without pom.xml.
 - image

### Testing done

```[tasklist]
### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue
```

### CODE REVIEW

**Review by kinow — COMMENTED** [review]
Hi @limejuny 

Thank you for the excellent description, screenshot, and neat code.

>I tested with snapshot version of scriptler-plugin, but I don't know the new version of scriptler dependency is mandatory. So I create commit without pom.xml.

It it not mandatory. I am not sure if we should merge and release it now, as it fixes an issue that is probably annoying to many users, or try to spend some more time writing a unit test for this.

I have to sort out my user credentials to release the plug-in, so we are not able to release it right now. So I will leave it open and try to write a unit test later if I have time, or decide later whether this should be merged without a test (not desired for any change).

If you have time to spend preparing a unit test for this it would be excellent, and make merging and releasing this a lot easier - but if not, you already did great sharing the fix!

Thanks

Bruno

**Comment by limejuny** [comment]
> Hi @limejuny
> 
> Thank you for the excellent description, screenshot, and neat code.
> 
> > I tested with snapshot version of scriptler-plugin, but I don't know the new version of scriptler dependency is mandatory. So I create commit without pom.xml.
> 
> It it not mandatory. I am not sure if we should merge and release it now, as it fixes an issue that is probably annoying to many users, or try to spend some more time writing a unit test for this.
> 
> I have to sort out my user credentials to release the plug-in, so we are not able to release it right now. So I will leave it open and try to write a unit test later if I have time, or decide later whether this should be merged without a test (not desired for any change).
> 
> If you have time to spend preparing a unit test for this it would be excellent, and make merging and releasing this a lot easier - but if not, you already did great sharing the fix!
> 
> Thanks
> 
> Bruno

Hello, @kinow 

I'd like to write a unit test, but I'm not familiar with Unit Testing, JUnit, or Jacoco, (or whatever).
Can you help me how to test and see coverage report (like ci.jenkins.io and `Files changed` tab in this page) on my local PC?
I want to write a test code (by referring to other test files), but I don't know how to check the test results.

Thanks.

**Comment by kinow** [comment]
Branch rebased.

>Can you help me how to test and see coverage report (like ci.jenkins.io and Files changed tab in this page) on my local PC?
>I want to write a test code (by referring to other test files), but I don't know how to check the test results.

The easiest way is by copying existing code. But first there must be an easy way to reproduce the bug. I will see if it's easy, and if I can write the test in <30-45 mins, otherwise maybe another day, or you can try to write it looking at the tests in the project :+1: 

Thanks!

Bruno

**Comment by kinow** [comment]
Huh, I couldn't find a way to reproduce it. The Job DSL example doesn't seem to be complete, I think? Or is it just for Scriptler? Could you provide a detailed step-by-step on how to reproduce this, please, @limejuny ? Also, could you confirm that this can only be merged after https://github.com/jenkinsci/scriptler-plugin/pull/80 is merged and released? Or can we test and release it before that's merged? Thanks!

**Comment by limejuny** [comment]
> Huh, I couldn't find a way to reproduce it. The Job DSL example doesn't seem to be complete, I think? Or is it just for Scriptler? Could you provide a detailed step-by-step on how to reproduce this, please, @limejuny ? Also, could you confirm that this can only be merged after [jenkinsci/scriptler-plugin#80](https://github.com/jenkinsci/scriptler-plugin/pull/80) is merged and released? Or can we test and release it before that's merged? Thanks!

I think this can be merge without scriptler's PR merge, but it will doesn't work.
In now, Job DSL can't use scriptler script, because `ScriptlerBuild` don't have `DataBoundConstructor` annotation, and the jenkinsci/scriptler-plugin#80 fix it.

To use scriptler script, 
1. the scriptler's PR should be merged
2. and new scriptler's version (which contain that fixed commit) released
3. **update dependency in this plugin's `pom.xml`** (that's why I use snapshot version dependency in comment's screenshot)

So, to reproduce this,
1. `wait for new version of scriptler` or `build the plugin with that PR locally`
2. and update the pom.xml, and build this plugin
3. install those two plugin and test
---
↑ After this, you can use scriptler script in Job DSL.
But even after that, scriptler script can be used, but parameters cannot be created by Job DSL.
That bug is [JENKINS-72105](https://issues.jenkins.io/browse/JENKINS-72105), and this PR will fix that.

**Comment by kinow** [comment]
Thanm you for the details,  @limejuny . I'll wait for scriptler to release so we can write a test and merge this one.

**Comment by mtughan** [comment]
@kinow, [Scriptler 334.v29792d5a_c058](https://github.com/jenkinsci/scriptler-plugin/releases/tag/334.v29792d5a_c058) has now been released with the fix. Please feel free to proceed.

**Comment by kinow** [comment]
> @kinow, [Scriptler 334.v29792d5a_c058](https://github.com/jenkinsci/scriptler-plugin/releases/tag/334.v29792d5a_c058) has now been released with the fix. Please feel free to proceed.

Thanks! I am overseas attending a $work conference for the rest of the week. Will leave the GitHub notification in my inbox to work on this later.

**Comment by limejuny** [comment]
Hi, 
I tested with `2.8.0` and this PR's plugin, with [this script](https://issues.jenkins.io/browse/JENKINS-72105).
and It worked, but something another bug appeared.

I used the scriptler script in `cascadeChoiceParameter`, but in `2.8.0`, there is no `cascadeChoiceParameter` in Job DSL Reference site. (<jenkins-url>/plugin/job-dsl/api-viewer/index.html)
(But it still work even if there is no `cascadeChoiceParameter` in reference.)
image

---
It seems to have nothing to do with this PR because it is already happening in version 2.8.0, but I would like to inform you because this may make the test difficult.

This is the more detailed Job DSL script.

---
```
pipelineJob("Test") {
    parameters {
        activeChoiceParam('PROJECT_NAME') {
            choiceType('SINGLE_SELECT')
            filterable()
            groovyScript {
                ~~~
            }
        }
        activeChoiceParam('CLOUD_HOST_TYPE') { ~~~ }
        activeChoiceParam('TEST_PROFILE') { ~~~ }
        cascadeChoiceParameter {
            name('SCRIPTLER_TEST')
            choiceType('PT_SINGLE_SELECT')
            referencedParameters('PROJECT_NAME,CLOUD_HOST_TYPE,TEST_PROFILE')
            script {
                scriptlerScript {
                    scriptlerBuilder {
                        builderId('org.jenkinsci.plugins.scriptler.builder.ScriptlerBuilder')
                        scriptId('TestImageTags.groovy')
                        propagateParams(true)
                        parameters {
                            parameter {
                                name 'PROJECT_NAME'
                                value '$PROJECT_NAME'
                            }
                            parameter {
                                name 'CLOUD_HOST_TYPE'
                                value '$CLOUD_HOST_TYPE'
                            }
                            parameter {
                                name 'PROFILE'
                                value '$TEST_PROFILE'
                            }
                        }
                        isSandboxed(true)
                    }
                }
            }
            randomName('')
            filterable(false)
            filterLength(1)
        }
    }

    definition {
        cps {
            ~~~
        }
    }
}
```

**Comment by kinow** [comment]
Thanks @limejuny 

I didn't have time to have a look at this issue over this weekend (returned home yesterday). But will try to finish reviewing next weekend.

About the missing cascade parameter, I wonder if that could be related to recent changes where we added the symbol marker for the job declarative form. We chose different names but not sure if that affects the same part of the UI in your screenshot - #43

**Comment by limejuny** [comment]
> Thanks @limejuny
> 
> I didn't have time to have a look at this issue over this weekend (returned home yesterday). But will try to finish reviewing next weekend.
> 
> About the missing cascade parameter, I wonder if that could be related to recent changes where we added the symbol marker for the job declarative form. We chose different names but not sure if that affects the same part of the UI in your screenshot - #43

Hi, @kinow 
Thank you for the hint, I think the `@Symbol` is cause.

When I revert [this](https://github.com/jenkinsci/active-choices-plugin/commit/36a8a567b71fefb93eedb6ff3624b1ea92f6cff5) and [this](https://github.com/jenkinsci/active-choices-plugin/commit/492eb593f8a4baf97ddd9b16660dea28df962831), the missing `cascadeChoiceParameter` and `dynamicReferenceParameter` turn back.
However, in the newest plugin, `activeChoice` and `activeChoiceHtml` are exists, but there are missed when revert those two commits.

(newest version)
image

(reverted version)
image

So, when I remove those `@Symbol`s in only `CascadeChoiceParameter.java` and `DynamicReferenceParameter.java`, the `activeChoice`, `cascadeChoiceParameter`, `dynamicReferenceParameter` seem well! (like this)
image

---

I don't know anything about the `@Symbol` annotation, so I don't know if I can revert those two files, and I think that doesn't seem to be directly involved this PR. I think that is another issue.

Thanks!

**Comment by kinow** [comment]
@limejuny what I will test next weekend (great if you can test that before, but you don't need to if you don't have time) is if the names we used for the `@Symbol` annotation created new links. For example, I **think** `cascadeChoiceParameter` would now be `activeChoice` (or another value configured in `@Symbol`).

So what I want to test is whether you can still use the plug-in like before but you must use the new names, or if that change removed some feature.

**Comment by limejuny** [comment]
> @limejuny what I will test next weekend (great if you can test that before, but you don't need to if you don't have time) is if the names we used for the `@Symbol` annotation created new links. For example, I **think** `cascadeChoiceParameter` would now be `activeChoice` (or another value configured in `@Symbol`).
> 
> So what I want to test is whether you can still use the plug-in like before but you must use the new names, or if that change removed some feature.

Oh, thank you. I couldn't think the `Symbol` would be that function!
I'll test it on Tuesday or Wednesday.
I hope the test results come out as expected.

Thanks.

**Comment by kinow** [comment]
Great @limejuny , if you can do that that will save my a lot of time before next weekend. Otherwise I can test that too, and report back in my review feedback (if any), before merging it. Thanks!

**Comment by limejuny** [comment]
I was just curious, so I tested it for a second, and it work!!

I just replace `cascadeChoiceParameter` to `reactiveChoice`.

image
(ref; [this](https://github.com/jenkinsci/active-choices-plugin/commit/492eb593f8a4baf97ddd9b16660dea28df962831#diff-130ec60325e13cfb6ba15297d03b7a510f9dc7153504be319f7f452476fd553a))

and this is the result.

#### uno-choice:2.8.0, same script above (without `cascadeChoiceParameter` -> `reactiveChoice`)
image

#### with this PR
image

---

There is no test case in this PR, so it may malfunction in parts other than what I tested. But it seems to work well, at least on Job DSL.

**Comment by kinow** [comment]
Thank you so much @limejuny !

>There is no test case in this PR, so it may malfunction in parts other than what I tested. But it seems to work well, at least on Job DSL.

:+1: all good. As you may have seen your commits have been tested by our current tests, and there were no failures. So this weekend I will sync the code (which is quite small, so it should be easier to review), and see how hard it would be to copy one of the existing tests and write one for the part of the code you changed.

From a quick look, it seems like you are making the code more defensive, which never hurts. I don't think it will break anything, but will do some testing (i.e. we should be able to merge and maybe release it over next weekend, so you can install this new version and play with that).

Should you find any bug we can always just revert and cut a new release :slightly_smiling_face:

**Comment by kinow** [comment]
Checked out the branch and am testing it now. Just set the breakpoint in the line changed, and am now trying to make that code to be called. Tried with a simple active choices parameter created manually with a scriptler script + parameter, but no luck. Will try the Job DSL now...

**Review by kinow — CHANGES_REQUESTED** [review]
Hi!

I tried:

- Declarative pipeline
- Pipeline
- Job DSL

None worked. Both issue and the comments in this pull request are lacking a minimal reproducible example. Without that I am not able to confirm that the issue occurs, nor test the code -- some issues include a config.xml that can be used as-is, or code snippets that can be copy-pasted without modifying anything, or simple step-by-step instructions on how to reproduce the issue.

I am replying in the JIRA issue too :+1:

**Comment by kinow** [comment]
> Hmm,,, and the CI is broken after rebase.

I think Jenkins infra changed something and either we now need a new setting in our pom.xml (a really common task when maintaining Jenkins plugins) or it's just a glitch on their end. Don't worry about fixing it for now, you've helped a lot now with the steps to reproduce the issue. I have more time today (bank holiday) so I will have another go. Thanks @limejuny !!! :bow:

**Comment by kinow** [comment]
Following @limejuny I was finally able to debug and hit the line changed in this pull request :partying_face: :tada: Thanks! Looking at the test now...

image

File: pom.xml
Comment (kinow) [inline pom.xml:104]: Fixes a CVE warning in the IDE.

File: pom.xml
Comment (kinow) [inline pom.xml:70]: Added this as the plug-in version was 2.387 (also needed for latest job-dsl I think, at least from the console logs).

File: pom.xml
Comment (kinow) [inline pom.xml:172]: Added these two to replicate the setup described by @limejuny

File: src/main/java/org/biouno/unochoice/model/ScriptlerScript.java
Comment (kinow) [inline src/main/java/org/biouno/unochoice/model/ScriptlerScript.java:132]: To fix deprecated warnings.

File: src/main/java/org/biouno/unochoice/model/ScriptlerScript.java
Comment (kinow) [inline src/main/java/org/biouno/unochoice/model/ScriptlerScript.java:272]: (`getParameters()` is annotated with `@NotNull`, just another change to make IDE happier)

Comment (kinow) [inline src/main/java/org/biouno/unochoice/model/ScriptlerScript.java:272]: (`getParameters()` is annotated with `@NotNull`, just another change to make IDE happier)

File: src/test/java/org/biouno/unochoice/issue72105/TestJobDslWithScriptlerScriptWithParameters.java
Comment (kinow) [inline src/test/java/org/biouno/unochoice/issue72105/TestJobDslWithScriptlerScriptWithParameters.java:110]: Not clear to my why UNSTABLE, but the job runs fine and produces the `scriptler-test` parameter (same locally :shrug:)

Comment (kinow) [inline src/test/java/org/biouno/unochoice/issue72105/TestJobDslWithScriptlerScriptWithParameters.java:110]: Not clear to my why UNSTABLE, but the job runs fine and produces the `scriptler-test` parameter (same locally :shrug:)

File: src/test/java/org/biouno/unochoice/issue72105/TestJobDslWithScriptlerScriptWithParameters.java
Comment (kinow) [inline src/test/java/org/biouno/unochoice/issue72105/TestJobDslWithScriptlerScriptWithParameters.java:139]: @limejuny , this is the test that I think confirms your change works and prevents us from breaking the feature/behavior of the plug-in again.

Basically, it's a normal JUnit test, annotated with `@Test` but also with `@LocalData`. This last annotation instructions Jenkins test harness to use a `config.xml` that I saved after reproducing your setup (I just copied a file like `/work/jobs/job/config.xml` into the test resources correct location).

This means that when the test is started, because it has the `@Rule JenkinsRule` object, a Jenkins instance is created for this test, and the same `job` you described in the issue is created by Jenkins for test. Then this test above was quickly written (could be simpler I think, but strapped for time now) just checks the Job DSL creates a new job, and finally confirms that the `P3` parameter contains the scriptler script with the parameters persisted.

I executed this with the IDE, while I still had the breakpoint in the line you changed, and that was triggered as expected.

File: src/test/resources/org/biouno/unochoice/issue72105/TestJobDslWithScriptlerScriptWithParameters/test/jobs/job/config.xml
Comment (kinow) [inline src/test/resources/org/biouno/unochoice/issue72105/TestJobDslWithScriptlerScriptWithParameters/test/jobs/job/config.xml:1]: File copied after I created the job following @limejuny 's instructions.

**Review by kinow — APPROVED** [review]
Test added. Just pending a green CI build, then it should be ready to be merged. Thanks @limejuny !

**Comment by kinow** [comment]
Build on `master` seems to be broken now too. Looks to me like something related to a new version of one of the plugin dependencies released, causing some conflicts in the tests :disappointed_relieved: Not sure where to find more information about it, so for now I think the best we can do is wait a little and see if there's a new release of jquery3-api or scriptler that fixes it. Sorry @limejuny

**Comment by kinow** [comment]
(And it does happen with some frequency, and I used to spend more time investigating these issues when I was using Jenkins professionally, but nowadays I am only using GH Actions and GitLab at $work, so completely out of the loop and without time to dig into Jenkins infra or dependency issues with other plugins & the latest LTS releases...)

File: src/test/java/org/biouno/unochoice/issue72105/TestJobDslWithScriptlerScriptWithParameters.java
Comment (kinow) [reply src/test/java/org/biouno/unochoice/issue72105/TestJobDslWithScriptlerScriptWithParameters.java]: Passing now after fixing master and rebasing here.

**Comment by kinow** [comment]
Fixed `master`, had to include the dependencies or a dependency :shrug: Merging. Thanks @limejuny !

**Comment by kinow** [comment]
Released with 2.8.1 :+1: (remember to test before moving to production, back up your server, etc.). Thanks for the pull request and patience @limejuny !

**Comment by limejuny** [comment]
> Released with 2.8.1 👍 (remember to test before moving to production, back up your server, etc.). Thanks for the pull request and patience @limejuny !

Thanks, I applied it to testing/production Jenkins instances. and it worked very well (at least, in my team's use case)!

**Pull request merged**
